### PR TITLE
heap working and all tests passing

### PR DESCRIPTION
**Pull request by paulsheridan**

### CODE REVIEW

**Comment by joelstanner** [comment]
Be sure to make frequent and atomic commits. This 1 commit has too much going on. A good rule of thumb is a new commit for a new function or a change on a function.

File: src/test_heap.py [hunk 1]
```diff
@@ -1,12 +1,52 @@
...
+import pytest
 
 
-TEST_LIST_ONE = [40, 33, 60, 19, 10]
-TEST_LIST_TWO = [34, 56, 78, 90, 3, 4, 1, 6, 29, 66, 5, 4, 8, 9, 24, 20]
-TEST_LIST_THREE = [24, 45, 92, 93, 65, 87, 19, 9, 46, 72, 85, 4, 76, 900]
+TEST_LISTS = ([40, 33, 60, 19, 10],
+              [34, 56, 78, 90, 9, 24, 20, 3, 4, 1, 6, 29, 66, 5, 4, 8],
+              [24, 45, 92, 93, 65, 87, 19, 9, 46, 72, 85, 4, 76, 900],
+              [34, 56, 78, 90, 3, 4, 1, 6, 29, 66, 5, 4, 8, 9, 24, 20, 24, 45, 92, 93, 65, 87, 19, 9, 46, 72, 85, 4, 76, 900])
```
Comment (joelstanner) [inline src/test_heap.py:8]: If possible, keep lines under 80 characters. See [python's implicit line joining](https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length).

Comment (joelstanner) [inline src/test_heap.py:8]: Remember to test edge cases. What happens when your list is only 1 node? What about an empty list?

File: src/heap.py [hunk 2]
```diff
@@ -12,28 +12,30 @@ def __init__(self, user_input=None):
...
+            print(idx)
+            parent = idx
+            child_l = idx * 2 + 1
+            child_r = idx * 2 + 2
+            idx += 1
+            print(self.heap)
+            if self.heap[parent] < self.heap[child_l]:
+                self.heap[parent], self.heap[child_l] = self.heap[child_l], self.heap[parent]
+            if self.heap[parent] < self.heap[child_r]:
+                self.heap[parent], self.heap[child_r] = self.heap[child_r], self.heap[parent]
```
Comment (joelstanner) [inline src/heap.py:37]: For this class, try and adhere to the PEP8 style guide and keep line lengths under 80.

File: src/heap.py [hunk 3]
```diff
@@ -12,28 +12,30 @@ def __init__(self, user_input=None):
...
+            continue
 
     def push(self, val):
```
Comment (joelstanner) [inline src/heap.py:40]: Make sure to add well formed docstrings to your methods.

File: src/heap.py [hunk 4]
```diff
@@ -43,15 +45,17 @@ def pop(self):
...
 
-new_heap = Heap([34, 56, 78, 90, 3, 4, 1, 6, 29, 66, 5, 4, 8, 9, 24, 20,
+
+new_heap = Heap([10, 7, 8, 3, 1, 4, 2, 12])
+big_heap = Heap([34, 56, 78, 90, 3, 4, 1, 6, 29, 66, 5, 4, 8, 9, 24, 20,
                  24, 45, 92, 93, 65, 87, 19, 9, 46, 72, 85, 4, 76, 900])
-print(new_heap.heap)
-new_heap.pop()
-print(new_heap.heap)
+print(big_heap.pop())
```
Comment (joelstanner) [inline src/heap.py:61]: Adding these test cases in the file should be done in an `if __name__ == '__main__':` block.

File: src/heap.py [hunk 5]
```diff
@@ -43,15 +45,17 @@ def pop(self):
         try:
             return_val = self.heap[0]
             del self.heap[0]
-            self.heap.insert(0, self.heap.pop(len(self.heap) - 1))
-            # self._sift_down()
+            self.heap.insert(0, self.heap.pop(self.length() - 1))
```
Comment (joelstanner) [inline src/heap.py:48]: Consider adding a `__len__` special method to your class. https://docs.python.org/3/reference/datamodel.html#object.__len__
